src/node_url.cc

```diff
@@ -799,117 +799,24 @@ bool ToASCII(const std::string& input, std::string* output) {
-  if (ch == ':') {
-    if (length < 2 || pointer[1] != ':')
-      return;
-    pointer += 2;
-    ch = pointer < end ? pointer[0] : kEOL;
-    piece_pointer++;
-    compress_pointer = piece_pointer;
-  }
-  while (ch != kEOL) {
-    if (piece_pointer >= buffer_end)
-      return;
-    if (ch == ':') {
-      if (compress_pointer != nullptr)
-        return;
-      pointer++;
-      ch = pointer < end ? pointer[0] : kEOL;
-      piece_pointer++;
-      compress_pointer = piece_pointer;
-      continue;
-    }
-    value = 0;
-    len = 0;
-    while (len < 4 && IsASCIIHexDigit(ch)) {
-      value = value * 0x10 + hex2bin(ch);
-      pointer++;
-      ch = pointer < end ? pointer[0] : kEOL;
-      len++;
-    }
-    switch (ch) {
-      case '.':
-        if (len == 0)
-          return;
-        pointer -= len;
-        ch = pointer < end ? pointer[0] : kEOL;
-        if (piece_pointer > buffer_end - 2)
-          return;
-        numbers_seen = 0;
-        while (ch != kEOL) {
-          value = 0xffffffff;
-          if (numbers_seen > 0) {
-            if (ch == '.' && numbers_seen < 4) {
-              pointer++;
-              ch = pointer < end ? pointer[0] : kEOL;
-            } else {
-              return;
-            }
-          }
-          if (!IsASCIIDigit(ch))
-            return;
-          while (IsASCIIDigit(ch)) {
-            unsigned number = ch - '0';
-            if (value == 0xffffffff) {
-              value = number;
-            } else if (value == 0) {
-              return;
-            } else {
-              value = value * 10 + number;
-            }
-            if (value > 255)
-              return;
-            pointer++;
-            ch = pointer < end ? pointer[0] : kEOL;
-          }
-          *piece_pointer = *piece_pointer * 0x100 + value;
-          numbers_seen++;
-          if (numbers_seen == 2 || numbers_seen == 4)
-            piece_pointer++;
-        }
-        if (numbers_seen != 4)
-          return;
-        continue;
-      case ':':
-        pointer++;
-        ch = pointer < end ? pointer[0] : kEOL;
-        if (ch == kEOL)
-          return;
-        break;
-      case kEOL:
-        break;
-      default:
-        return;
-    }
-    *piece_pointer = value;
-    piece_pointer++;
-  }
 
-  if (compress_pointer != nullptr) {
-    int64_t swaps = piece_pointer - compress_pointer;
-    piece_pointer = buffer_end - 1;
-    while (piece_pointer != &value_.ipv6[0] && swaps > 0) {
-      uint16_t temp = *piece_pointer;
-      uint16_t* swap_piece = compress_pointer + swaps - 1;
-      *piece_pointer = *swap_piece;
-      *swap_piece = temp;
-       piece_pointer--;
-       swaps--;
-    }
-  } else if (compress_pointer == nullptr &&
-             piece_pointer != buffer_end) {
+  unsigned char buf[sizeof(struct in6_addr)];
+  memset(buf, 0, sizeof(buf));
+
+  char* zero = const_cast<char*>(input + length);
+  char origin_zero_char = *zero;
+  *zero = 0;
```

Comment:
`const_cast` is for cases in which you know that the resulting pointer will not be modified, even though it could be, *not* for invalidating the guarantees made by `const`. In particular, even though this resets the value to its original value, it does not enforce that `input` is actually located in a writable part of the memory.

I think you’ll just need to copy `input` in this case.

Reply:
I found input is a string's c_str, so I changed the title to WIP.

You're right, I should copy that memoty.